File: 2025/5/main.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define LINE_SIZE 1024
#define MAX_RANGES 200
#define MAX_INGREDIENTS 1000
typedef long long IngredientId;

typedef struct{
    IngredientId start;
    IngredientId end;
} Range;
typedef Range Ranges[MAX_RANGES];

int valid(Range* range) {
  return range->start != -1 && range->end != -1;
}

int compare_ranges(const void *lhs, const void *rhs) {
  Range* lhs_range = (Range *)lhs;
  Range* rhs_range = (Range *)rhs;
  const int lhs_valid = valid(lhs_range);
  const int rhs_valid = valid(rhs_range);

  // Put invalid ranges at the end
  if (! lhs_valid && ! rhs_valid) {
    return 0;
  }
  if (! lhs_valid) {
    return 1;
  }
  if (! rhs_valid) {
    return -1;
  }

  if (lhs_range->start < rhs_range->start) {
    return -1;
  }
  if (lhs_range->start > rhs_range->start) {
    return 1;
  }
  return 0;
}

long long solve1(Ranges ranges, IngredientId ingredients[MAX_INGREDIENTS]) {
  long long result = 0;
  int fresh_lines = 1;
  for (int i = 0; i < MAX_INGREDIENTS; i++) {
    IngredientId ingredient_id = ingredients[i];
    if (ingredient_id == -1) {
      break;
    }
    for (int i = 0; i < MAX_RANGES; i++) {
      Range range = ranges[i];
      if (range.start <= ingredient_id && ingredient_id <= range.end) {
        result += 1;
        break;
      }
    }
  }
  return result;
}

long long solve2(Ranges ranges) {
  long long result = 0;
  Ranges normal_ranges;
  for (int i = 0; i < MAX_RANGES; i++) {
      normal_ranges[i].start = -1;
      normal_ranges[i].end = -1;
  }
  for (int i = 0; i < MAX_RANGES; i++) {
    Range range = ranges[i];
    IngredientId start = range.start;
    IngredientId end = range.end;
    if (! valid(&range)) {
      break;
    }
    int start_included_in = -1;
    int end_included_in = -1;
    for (int n = 0; valid(&normal_ranges[n]); n++) {
      Range range_j = normal_ranges[n];
      IngredientId normal_start = range_j.start;
      IngredientId normal_end = range_j.end;
      if (normal_start <= start && start <= normal_end + 1) {
        start_included_in = n;
      }
      if (normal_start - 1 <= end && end <= normal_end) {
        end_included_in = n;
      }
      if (start_included_in >= 0 && start_included_in == end_included_in) {
        break;
      }
      if (start < normal_start && normal_end < end) {
        normal_ranges[n].start = -1;
        normal_ranges[n].end = -1;
      }
      if (start_included_in >= 0 && end_included_in >= 0) {
        break;
      }
      if (start < normal_end && end < normal_end) {
        break;
      }
    }
    if (start_included_in == -1 && end_included_in == -1) {
      int first_invalid_index = 0;
      while (valid(&normal_ranges[first_invalid_index])) {
        first_invalid_index++;
      };
      normal_ranges[first_invalid_index].start = start;
      normal_ranges[first_invalid_index].end = end;
      // print(f"Range {start}-{end} intersects nothing")
      printf("Range %lld-%lld intersects nothing\n",
          start,
          end);
    }
    else if (start_included_in == end_included_in) {
      printf("Range %lld-%lld is eclipsed by %lld-%lld\n",
          start,
          end,
          normal_ranges[start_included_in].start,
          normal_ranges[start_included_in].end);
      continue;
    }
    else if (start_included_in != -1 && end_included_in == -1) {
      printf("Range %lld-%lld intersects %lld-%lld at start\n",
          start,
          end,
          normal_ranges[start_included_in].start,
          normal_ranges[start_included_in].end);
      normal_ranges[start_included_in].end = end;
      printf("Extending to %lld-%lld\n",
          normal_ranges[start_included_in].start,
          normal_ranges[start_included_in].end);
    }
    else if (start_included_in == -1 && end_included_in != -1) {
      printf("Range %lld-%lld intersects %lld-%lld at end\n",
          start,
          end,
          normal_ranges[end_included_in].start,
          normal_ranges[end_included_in].end);
      normal_ranges[end_included_in].start = start;
      printf("Extending to %lld-%lld\n",
          normal_ranges[end_included_in].start,
          normal_ranges[end_included_in].end);
    }
    else {
        printf("Range %lld-%lld intersects %lld-%lld and %lld-%lld\n",
          start,
          end,
          normal_ranges[start_included_in].start,
          normal_ranges[start_included_in].end,
          normal_ranges[end_included_in].start,
          normal_ranges[end_included_in].end);
        normal_ranges[start_included_in].end = normal_ranges[end_included_in].end;
        normal_ranges[end_included_in].start = -1;
    }
    qsort(normal_ranges, MAX_RANGES, sizeof(Range), compare_ranges);
    for (int n = 0; valid(&normal_ranges[n]); ++n) {
      Range range_j = normal_ranges[n];
    }
  }

  for (int i = 0; valid(&normal_ranges[i]); ++i) {
    Range normal_range = normal_ranges[i];
    printf("  Normal range %d: %lld-%lld\n", i, normal_range.start, normal_range.end);
    result += normal_range.end - normal_range.start + 1;
  }
  return result;
}

int main () {
  Ranges fresh_ranges;
  for (int i = 0; i < MAX_RANGES; i++) {
      fresh_ranges[i].start = -1;
      fresh_ranges[i].end = -1;
  }
  char buffer[LINE_SIZE];
  IngredientId ingredients[MAX_INGREDIENTS];
  for (int i = 0; i < MAX_INGREDIENTS; i++) {
    ingredients[i] = -1;
  }
  int fresh_lines = 1;
  int i = 0;
  while (fgets(buffer, LINE_SIZE, stdin)) {
    if (strlen(buffer) <= 1) {
      fresh_lines = 0;
      i = 0;
      continue;
    }
    if (fresh_lines) {
      sscanf(buffer, "%lld-%lld", &fresh_ranges[i].start, &fresh_ranges[i].end);
      i++;
    }
    else {
      ingredients[i] = atoll(buffer);
      i++;
    }
  }
  const long long solution1 = solve1(fresh_ranges, ingredients);
  if (solution1 != 0) {
    printf("Solution 1: %lld\n", solution1);
  }
  const long long solution2 = solve2(fresh_ranges);
  if (solution2 != 0) {
    printf("Solution 2: %lld\n", solution2);
  }
}

File: 2025/5/main.py
#!/usr/bin/env python3
import sys
sys.path.append('../..')
from py_utils.grid import *
from py_utils.fast_parse import *
from py_utils.search import *

from collections import defaultdict

import fileinput
import itertools
import math
import os
import queue
import re

def dprint(*args, **kwargs):
    if False:
        return print(*args, **kwargs)

p1 = None
p2 = None

def solve1(lines):
    r = 0
    fresh_lines = True;
    fresh_ranges = []
    ingredient_ids = []
    for line in lines:
        if line == '':
            fresh_lines = False
            continue
        if fresh_lines:
            start, end = line.split('-')
            fresh_ranges.append((start, end))
        else:
            ingredient_ids.append(int(line))
    for ingredient_id in ingredient_ids:
        for start, end in fresh_ranges:
            if int(start) <= ingredient_id <= int(end):
                r += 1
                break
    return r

def solve2(lines):
    r = 0
    fresh_ranges = []
    ingredient_ids = []
    for line in lines:
        if line == '':
            break
        start, end = [int(num) for num in line.split('-')]
        assert(start <= end)
        fresh_ranges.append((start, end))
    normal_ranges = []
    for start, end in fresh_ranges:
        start_included_in = None
        end_included_in = None
        for i in range(len(normal_ranges)):
            range_i = normal_ranges[i]
            normal_start, normal_end = range_i
            if normal_start <= start <= normal_end + 1:
                start_included_in = i
                dprint(f"{start} is in {normal_start}-{normal_end}")
            else:
                dprint(f"{start} not in {normal_start}-{normal_end}")
            if normal_start - 1 <= end <= normal_end:
                dprint(f"{end} is in {normal_start}-{normal_end}")
                end_included_in = i
            else:
                dprint(f"{end} not in {normal_start}-{normal_end}")
            if start_included_in is not None and start_included_in == end_included_in:
                dprint(f"Range {start}-{end} is fully included in {normal_ranges[i]}")
                break
            if start < normal_start and normal_end < end:
                dprint("Removing", normal_ranges[i])
                normal_ranges[i] = (0, 0)
            if start_included_in is not None and end_included_in is not None:
                break
            if start < normal_end and end < normal_end:
                break
        if start_included_in is None and end_included_in is None:
            dprint(f"Range {start}-{end} intersects nothing")
            normal_ranges.append((start, end))
        elif start_included_in == end_included_in:
            dprint(f"Range {start}-{end} is eclipsed by {normal_ranges[start_included_in]}")
            continue
        elif start_included_in is not None and end_included_in is None:
            dprint(f"Range {start}-{end} intersects {normal_ranges[start_included_in]}")
            normal_ranges[start_included_in] = (
                normal_ranges[start_included_in][0],
                end
            )
            dprint(f"Extending existing range to {normal_ranges[start_included_in]}")
        elif start_included_in is None and end_included_in is not None:
            dprint(f"Range {start}-{end} intersects {normal_ranges[end_included_in]}")
            normal_ranges[end_included_in] = (
                start,
                normal_ranges[end_included_in][1]
            )
            dprint(f"Extending existing range to {normal_ranges[end_included_in]}")
        else:
            dprint(f"Range {start}-{end} intersects {normal_ranges[start_included_in]} and {normal_ranges[end_included_in]}")
            normal_ranges[start_included_in] = (
                normal_ranges[start_included_in][0],
                normal_ranges[end_included_in][1]
            )
            dprint(f"Removing {normal_ranges[end_included_in]}")
            normal_ranges[end_included_in] = (0, 0)
        normal_ranges = [r for r in normal_ranges if r != (0, 0)]
        for normal_range in normal_ranges:
            assert(start <= end)
        normal_ranges = sorted(normal_ranges)
    prev_num = 0
    dprint(normal_ranges)
    for index, normal_range in enumerate(normal_ranges):
        r += normal_range[1] - normal_range[0] + 1
        print(f"  Normal range {index}: {normal_range[0]}-{normal_range[1]}");
        if normal_range[0] <= prev_num:
            dprint("Overlap!", normal_range[0])
        if normal_range[1] <= prev_num:
            dprint("Overlap!", normal_range[1])
        prev_num = normal_range[1]
    return r

if __name__ == "__main__":
    f = fileinput.input()
    lines = []
    for line in f:
        if line[-1] == '\n':
            line = line[:-1]
        lines.append(line)
    p1 = solve1(lines)
    p2 = solve2(lines)
    if p1 is not None:
        print("Solution 1:", p1)
    if p2 is not None:
        print("Solution 2:", p2)

   # generate a simplified input
   # for line in lines:
   #     factor = 1000000000000
   #     if line == '':
   #         print(line)
   #         continue
   #     if '-' not in line:
   #         print(int(line) // factor)
   #     else:
   #         start, end = line.split('-')
   #         print(f'{int(start) // factor}-{int(end) // factor}')
